File: backend/mr-db-api/src/main/java/io/featurehub/db/api/RolloutStrategyValidator.java
package io.featurehub.db.api;

import io.featurehub.mr.model.RolloutStrategy;
import io.featurehub.mr.model.RolloutStrategyCollectionViolationType;
import io.featurehub.mr.model.RolloutStrategyInstance;
import io.featurehub.mr.model.RolloutStrategyViolation;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public interface RolloutStrategyValidator {

  class ValidationFailure {
    public Map<RolloutStrategy, Set<RolloutStrategyViolation>> customStrategyViolations = new HashMap<>();
    public Map<RolloutStrategyInstance, Set<RolloutStrategyViolation>> sharedStrategyViolations = new HashMap<>();
    public Set<RolloutStrategyCollectionViolationType> collectionViolationType = new HashSet<>();

    public void add(RolloutStrategyViolation failure, RolloutStrategy strategy) {
      customStrategyViolations.computeIfAbsent(strategy, (k) -> new HashSet<>()).add(failure);
    }

    public void add(RolloutStrategyViolation failure, RolloutStrategyInstance rsi) {
      sharedStrategyViolations.computeIfAbsent(rsi, (k) -> new HashSet<>()).add(failure);
    }

    public void add(RolloutStrategyCollectionViolationType failure) {
      collectionViolationType.add(failure);
    }

    public void hasFailedValidation() throws InvalidStrategyCombination {
      if (isInvalid()) {
        throw new InvalidStrategyCombination(this);
      }
    }

    public boolean isInvalid() {
      return !(customStrategyViolations.isEmpty() && sharedStrategyViolations.isEmpty() && collectionViolationType.isEmpty());
    }
  }

  class InvalidStrategyCombination extends Exception {
    public final ValidationFailure failure;

    public InvalidStrategyCombination(ValidationFailure failure) {
      this.failure = failure;
    }
  }

  ValidationFailure validateStrategies(List<RolloutStrategy> customStrategies,
                             List<RolloutStrategyInstance> sharedStrategies);
  ValidationFailure validateStrategies(List<RolloutStrategy> customStrategies,
                             List<RolloutStrategyInstance> sharedStrategies, ValidationFailure failure);
}

File: backend/mr-db-api/src/main/kotlin/io/featurehub/db/api/RolloutStrategyValidator.kt
package io.featurehub.db.api

import io.featurehub.mr.model.*

interface RolloutStrategyValidator {
  class ValidationFailure {
    val customStrategyViolations = mutableMapOf<RolloutStrategy, MutableSet<RolloutStrategyViolation>>()
    val sharedStrategyViolations = mutableMapOf<RolloutStrategyInstance, MutableSet<RolloutStrategyViolation>>()
    val collectionViolationType = mutableSetOf<RolloutStrategyCollectionViolationType>()

    fun add(failure: RolloutStrategyViolation, strategy: RolloutStrategy) {
      customStrategyViolations.computeIfAbsent(strategy) { mutableSetOf() }.add(failure)
    }

    fun add(failure: RolloutStrategyViolation, rsi: RolloutStrategyInstance) {
      sharedStrategyViolations.computeIfAbsent(rsi) { mutableSetOf() }.add(failure)
    }

    fun add(failure: RolloutStrategyCollectionViolationType) {
      collectionViolationType.add(failure)
    }

    @Throws(InvalidStrategyCombination::class)
    fun hasFailedValidation() {
      if (isInvalid) {
        throw InvalidStrategyCombination(this)
      }
    }

    val isInvalid: Boolean
      get() = !(customStrategyViolations.isEmpty() && sharedStrategyViolations.isEmpty() && collectionViolationType.isEmpty())
  }

  class InvalidStrategyCombination(val failure: ValidationFailure) : Exception()

  fun validateStrategies(
    featureValueType: FeatureValueType?, customStrategies: List<RolloutStrategy>,
    sharedStrategies: List<RolloutStrategyInstance>
  ): ValidationFailure

  fun validateStrategies(
    featureValueType: FeatureValueType?,
    customStrategies: List<RolloutStrategy>,
    sharedStrategies: List<RolloutStrategyInstance>, failure: ValidationFailure?): ValidationFailure
}
